File: services/bff/src/main/java/io/wkrzywiec/fooddelivery/bff/view/create/RedisOrdersChannelConsumer.java
package io.wkrzywiec.fooddelivery.bff.view.create;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.wkrzywiec.fooddelivery.bff.view.create.incoming.*;
import io.wkrzywiec.fooddelivery.commons.event.DomainMessageBody;
import io.wkrzywiec.fooddelivery.commons.infra.messaging.Header;
import io.wkrzywiec.fooddelivery.commons.infra.messaging.redis.RedisStreamListener;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.connection.stream.MapRecord;

@Slf4j
@RequiredArgsConstructor
public class RedisOrdersChannelConsumer implements RedisStreamListener {

    private final RedisDeliveryViewProcessor processor;
    private final ObjectMapper objectMapper;

    @Override
    public String streamName() {
        return "orders";
    }

    @Override
    public String group() {
        return "bff";
    }

    @Override
    public String consumer() {
        return "1";
    }

    @Override
    public void onMessage(MapRecord<String, String, String> message) {
        log.info("Message received from {} stream: {}", streamName(), message);

        var payloadMessage = message.getValue().get("payload");

        try {
            var messageAsJson = objectMapper.readTree(payloadMessage);
            Header header = map(messageAsJson.get("header"), Header.class);

            DomainMessageBody event = switch (header.type()) {
                case "DeliveryCreated" -> mapMessageBody(messageAsJson, DeliveryCreated.class);
                case "TipAddedToDelivery" -> mapMessageBody(messageAsJson, TipAddedToDelivery.class);
                case "DeliveryCanceled" -> mapMessageBody(messageAsJson, DeliveryCanceled.class);
                case "FoodInPreparation" -> mapMessageBody(messageAsJson, FoodInPreparation.class);
                case "DeliveryManAssigned" -> mapMessageBody(messageAsJson, DeliveryManAssigned.class);
                case "DeliveryManUnAssigned" -> mapMessageBody(messageAsJson, DeliveryManUnAssigned.class);
                case "FoodIsReady" -> mapMessageBody(messageAsJson, FoodIsReady.class);
                case "FoodWasPickedUp" -> mapMessageBody(messageAsJson, FoodWasPickedUp.class);
                case "FoodDelivered" -> mapMessageBody(messageAsJson, FoodDelivered.class);
                default -> null;
            };

            if (event == null) {
                log.info("There is no delivery view handling logic for {} message.", header.type());
                return;
            }
            processor.handle(event);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    private <T> T mapMessageBody(JsonNode fullMessage, Class<T> valueType) throws JsonProcessingException {
        return objectMapper.treeToValue(fullMessage.get("body"), valueType);
    }

    private <T> T map(JsonNode fullMessage, Class<T> valueType) throws JsonProcessingException {
        return objectMapper.treeToValue(fullMessage, valueType);
    }
}

File: services/ordering/src/main/kotlin/io/wkrzywiec/fooddelivery/ordering/infra/stream/RedisOrdersChannelConsumer.kt
package io.wkrzywiec.fooddelivery.ordering.infra.stream

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import io.wkrzywiec.fooddelivery.commons.infra.messaging.Header
import io.wkrzywiec.fooddelivery.commons.infra.messaging.redis.RedisStreamListener
import io.wkrzywiec.fooddelivery.commons.model.AddTip
import io.wkrzywiec.fooddelivery.commons.model.CancelOrder
import io.wkrzywiec.fooddelivery.commons.model.CreateOrder
import io.wkrzywiec.fooddelivery.ordering.domain.OrderingFacade
import io.wkrzywiec.fooddelivery.ordering.domain.incoming.FoodDelivered
import io.wkrzywiec.fooddelivery.ordering.domain.incoming.FoodInPreparation
import lombok.RequiredArgsConstructor
import lombok.extern.slf4j.Slf4j
import org.springframework.data.redis.connection.stream.MapRecord

@Slf4j
@RequiredArgsConstructor
class RedisOrdersChannelConsumer : RedisStreamListener {
    private val facade: OrderingFacade? = null
    private val objectMapper: ObjectMapper? = null

    override fun streamName(): String {
        return "orders"
    }

    override fun group(): String {
        return "ordering"
    }

    override fun consumer(): String {
        return "1"
    }

    override fun onMessage(message: MapRecord<String?, String?, String?>) {
        RedisOrdersChannelConsumer.log.info("Message received from {} stream: {}", streamName(), message)

        val payloadMessage = message.value["payload"]

        try {
            val messageAsJson = objectMapper!!.readTree(payloadMessage)
            val header = map(messageAsJson["header"], Header::class.java)

            when (header.type) {
                "CreateOrder" -> facade!!.handle(mapMessageBody(messageAsJson, CreateOrder::class.java))
                "CancelOrder" -> facade!!.handle(mapMessageBody(messageAsJson, CancelOrder::class.java))
                "FoodInPreparation" -> facade!!.handle(mapMessageBody(messageAsJson, FoodInPreparation::class.java))
                "AddTip" -> facade!!.handle(mapMessageBody(messageAsJson, AddTip::class.java))
                "FoodDelivered" -> facade!!.handle(mapMessageBody(messageAsJson, FoodDelivered::class.java))
                else -> RedisOrdersChannelConsumer.log.info("There is no logic for handling {} message", header.type)
            }
        } catch (e: JsonProcessingException) {
            throw RuntimeException(e)
        }
    }

    @Throws(JsonProcessingException::class)
    private fun <T> mapMessageBody(fullMessage: JsonNode, valueType: Class<T>): T {
        return objectMapper!!.treeToValue(fullMessage["body"], valueType)
    }

    @Throws(JsonProcessingException::class)
    private fun <T> map(fullMessage: JsonNode, valueType: Class<T>): T {
        return objectMapper!!.treeToValue(fullMessage, valueType)
    }
}
